### PR TITLE
perf(lite): adopt Core fusions and main_grad accumulation on the DSv4 path

### PR DESCRIPTION
**Pull request by ISEEKYAN**
- [ ] I, the PR author, have personally reviewed every line of this PR.

# What does this PR do?

Brings DeepSeek-V4-Flash on Megatron-Lite to step-time parity with the same
model on Megatron Core, by adopting fusions Core already ships and dropping costs
Core does not pay.

None of this is a new kernel. Each change either replaces a hand-written
implementation with an existing Core fusion, or removes work Core's own DSv4 path
never does. Production code is +354/-183 across five files; the rest is tests.

## Measurements

8xH100, DS4-Flash 8 layers / 64 experts / topk 6 / seq 4096 / gbs 32,
tp1 pp1 cp1 ep8, recompute full, bf16, distributed optimizer.

Both arms run the same configuration -- indexer loss off, no forced load
balancing. That is worth stating because the first comparisons here were not
like-for-like in either direction, and correcting them moved Core's number by 3%
in one case and 2% in the other.

| | ms/step (median) | runs | spread |
|---|---|---|---|
| Megatron Core | 1359.9 | 3 | 0.15% |
| **lite, this branch** | **1357.8** | 6 | 3.47% |

**Parity, not a speedup.** The medians differ by 0.15%, and this branch's own
run-to-run spread is an order of magnitude larger than that: the six samples are
1350.4, 1353.6, 1354.8, 1360.8, 1367.2, 1397.6 ms. Three of them are individually
slower than Core. The honest reading is that the two are indistinguishable at
this sample size, and the claim being made is that lite no longer costs 19% more,
not that it is faster.

The dispersion itself is unexplained and is recorded as a known issue below.
Core's arm, measured in the same session on the same nodes, held 0.15%.

The same harness on this branch's merge-base, with this branch's changes absent,
measures 2474.1 ms (single run). That figure is what these changes move.

Earlier measurements on the pre-#7064 base -- Core 1376.3, lite before 1595.3,
lite after 1368.1, and an FSDP2 pair at 2223.5 -> 1624.5 ms with peak memory
41.15 -> 40.13 GB -- are left here for the per-change attribution below, but they
predate the rebase and should not be read as current.

Per change, measured on its own on that earlier base, in wall-clock terms:

| | ms |
|---|---|
| CSA output projected in THD, no BSHD round trip | -142.4 |
| weight gradients accumulated into `main_grad` (four sites) | -45.9 |
| dist_opt collectives overlapped; untrainable indexer frozen | -24.1 |
| rotary tables built once and shared | -11.2 |

### CSA output projection stays in THD

The THD attention output is `(total, np, hn)`. Reaching the shared BSHD output
projection meant a `permute` into `(1, np, total, hn)`, an unfused inverse RoPE
that concatenated the untouched nope channels back onto the rotated ones, and a
`transpose` + `reshape` on the way out. Three full materialisations of the
largest tensor in the module, per layer, paid twice again under activation
recompute — a per-module NVTX profile attributed 64% of that module's kernel
time to `direct_copy` and `CatArrayBatchedCopy`.

None of it is needed. `(total, np, hn)` is already
`(total, o_groups, heads_per_group * hn)` by a plain view, and Core's fused MLA
RoPE undoes the rotation over the RoPE channels alone. This is the route Core's
own DSv4 attention takes — `fused_mla_rope_out_of_place(..., inverse=True)`,
out-of-place because fused DSA backward retains the raw attention output.

### Weight gradients accumulate inside the GEMM

Every weight gradient was materialised in full and then folded into the fp32
accumulator one parameter at a time:

```python
param.main_grad.add_(param.grad.data)   # distributed_data_parallel.py
```

An op census by tensor shape put the vocabulary projection, the embedding table
and the eight CSA projections at 46 ms per step between them, all on the stream
the collectives share. Core pays none of it: it passes
`fuse_wgrad_accumulation` into every TE linear, so the wgrad GEMM accumulates
into `main_grad` in place. The string did not appear anywhere in the lite tree.

Four sites, three mechanisms:

- **TE linears** — the flag is flipped after the DDP wrap, since lite's modules
  are built before `main_grad` exists. Self-gating: on where the buffer is,
  off for optimizers that allocate none.
- **The vocabulary projection** — calls Core's
  `linear_with_grad_accumulation_and_async_allreduce` rather than reproducing
  it. The handshake it performs — writing through `fused_weight_gradient_mlp`
  and still returning a gradient so DDP's post-hook fires — is easy to get
  subtly wrong, and an earlier hand-rolled version of exactly this was discarded
  for that reason.
- **The embedding table** — `index_add_` straight into the fp32 accumulator,
  which also removes the dense `[padded_vocab, hidden]` temporary. Core has no
  equivalent for its own embedding, so this one is lite's; the mechanism is
  still Core's, down to returning a placeholder gradient rather than `None`
  because DDP asserts one is present when `overlap_grad_reduce` is set.

### Collectives overlap, and the indexer is honest about being untrained

`overlap_grad_reduce` was hardcoded `False` and `overlap_param_gather` was never
set, so the collectives sat on the critical path.

Turning them on hit
`assert len(per_param_grad_ready_counts) == len(params)`, and the six parameters
per layer that never reported a gradient were the CSA indexer's — not the MoE
experts', as first assumed. The indexer runs on detached inputs and only selects
top-k, which is not differentiable, so those weights are trained by the indexer
loss and by nothing else. With that loss off they received no gradient ever,
while still occupying gradient buckets and optimizer state. They are frozen when
the coefficient is zero and unfrozen when it is not.

### Rotary tables are built once and indexed by the kernel

Every layer derived the same positions from the same `cu_seqlens` and rank
offset, so the tables were rebuilt identically eight times over, and again under
recompute.

Getting this right took four attempts, and the measurements are the argument:

```
cached globally against the position tensor's address    unsound
strong reference to that tensor, keeping it alive        1387.7 ms
weak reference, identity checked on hit                  1385.1 ms
no cache at all                                          1381.0 ms
attached to the packed-batch object                      1365.8 ms
Core's shape, rows gathered per layer in Python          1375.4 ms
Core's shape, table handed to the kernel                 1368.1 ms
```

The address-keyed version is the one review caught: the caching allocator hands
the same address to a later tensor of the same shape, so a lookup could return
tables built for entirely different positions — wrong RoPE, silently, with every
shape and dtype intact.

What shipped is Core's shape. The tables depend on the rope parameters and how
far the positions run, never on which tensor carries them, so the cache is keyed
on those scalars alone, grows when a longer span is asked for, and is indexed by
`cu_seqlens` inside the fused kernels — which the three calls here already pass.
Nothing depends on a tensor's identity or contents, so there is no version
counter, no liveness trick, and no way for an in-place edit to return a stale
table.

The intermediate step is worth naming because it looks like the same fix and is
not: building one table and then selecting each layer's rows in Python moved the
same bytes as the per-layer rebuild, and cost 10 ms against the object-attached
cache it replaced. Core's shape only pays off with Core's consumer. The one
caller that genuinely needs per-row tables — the indexer, through
`apply_partial_rope` — still gathers, and is now the only place that does.

The object-attached version was 2.3 ms faster, inside the noise floor. It is not
what shipped, because it still answered "which object is this" when the question
should never have been asked.

### Smaller

Clamped SwiGLU reaches `bias_swiglu_impl` / `weighted_bias_swiglu_impl` instead
of an eager chunk/clamp/silu/multiply in fp32 on the largest activation in the
model — only the unclamped variant had been fused, and DeepSeek-V4 always sets a
clamp. The mHC residual, aggregate and Sinkhorn each had a hand-written form
sitting next to a Core kernel computing the same expression. CSA builds q/kv in
SBHD and uses Core's fused MLA RoPE.

## Not supported

**Expert tensor parallelism.** The MoE experts had an ETP path that had never
been exercised, and its all-reduce disagreed with the one in `lora.py` about
whether the backward needs a second reduction. Both are correct under opposite
assumptions about whether the incoming gradient is already replicated, and
nothing recorded which one holds here; meanwhile the docstring claimed the
reduction the code does not perform. A user setting `etp_size > 1` got no error
and a gradient that may or may not have been scaled by the group size. It now
raises at construction.

## Tried and reverted

Recorded because the measurements are the useful part:

- **MoE permutation fusion, unconditionally.** Measured nothing, and revealed a
  behavioural difference rather than introducing one: with a token routed to the
  same expert twice, the fused permutation does not sum the duplicate scores the
  way the unfused one does, which four existing tests assert on. The environment
  variable defaulted to off, so that path had never been exercised. The
  difference predates this branch and is left to be settled separately;
  `dispatcher.py` is untouched here.
- **Indexing the all-to-all reshuffle.** Moved the same bytes. Zero.
- **Summing the mHC residual branch gradients in one fused add** via Core's
  `BroadcastTensorFused`. −0.18%, inside the noise floor.
- **Fusing `apply_partial_rope`.** +2.37%.
- **Gathering rotary rows per layer from a shared table.** +9.6 ms against the
  cache it replaced; see above.
- **Screening dense batches for CSA in the batch builder.** Put CSA's constraint
  in a layer with no business knowing it, and paid a module scan per forward.
  CSA's own error message now names the shape it got and the packing that works.

## Evidence

Each semantic change was adjudicated on GPU against the expression it replaces,
on the packed THD path production uses, with a negative control that must fail.
The inverse RoPE, for example:

```
fused(inverse) o fused(forward) vs identity   rel 5.59e-3
apply_partial_rope(cos, -sin) o fused vs id   rel 1.12e-2
new vs old, same input                        rel 5.56e-3
inverse=False vs identity  (control)          rel 1.62e+0
```

The replacement is the more accurate of the two: it rotates once in the kernel
instead of round-tripping bf16 through a rotate-and-concatenate.

The rotary change was adjudicated the same way, against the per-row tables it
replaces, through the same fused kernel:

```
whole table indexed by cu_seqlens vs per-row table   rel 0.000e+00
table shifted one position          (control)        rel 7.273e-01
```

A single-GPU `pytest` over the whole tree reports 5 failed, 809 passed, 24
skipped, 0 errors. The five failures are the same five that pristine `dev`
produces, item for item, and each carries its own explanation: four assert
`EP2/TP2 real-device tests require exactly two torchrun ranks, got 1`, and the
fifth spawns two gloo processes. They are rank-provisioning, not defects, and
belong to the repository's own runner rather than a single-GPU pytest. That
runner was not re-run here: it launches its workers with `sys.executable`, and
in this environment pytest reaches the interpreter through `PYTHONPATH` rather
than being installed, so the workers cannot import it. That is a property of the
environment, not of this branch.

End-to-end, the tree before these commits and after were each run three times on
the same data. The judge is not the loss — two runs of the same tree with the
same seed differ by 0.006. The noise floor is measured by re-running identical
code, and the cross-arm difference compared against it:

```
grad_norm   intra-arm spread 2.09e-4   cross-arm difference 7.9e-5   (0.38x)
loss        intra-arm spread 1.40e-2   cross-arm difference 8.0e-3   (0.57x)
```

The end-to-end pair predates the final rotary rewrite; that change is covered by
the bitwise adjudication above rather than by a fresh pair.

## Release notes

- `_per_head_rms` is a local fusion, not an upstream one, and is not
  bitwise-neutral: the compiler reassociates the mean, reaching one bf16 ulp.
  Bounded by a unit test and kept for a measured 8.75%. If Core grows an
  equivalent this should converge on it.
- `_project_context` is now reached only by the BSHD CP=1 routes; THD projects
  its output in place.
- The CPU forms of the fused mHC calls are gone. Production is always CUDA, and
  only tests reached them.
- `_enable_wgrad_accumulation_fusion` is covered by an import contract only; the
  flag it flips is exercised end-to-end but has no unit test of its own.
- The combination of distributed-optimizer wgrad writing `main_grad` with the
  expert TP `register_hook` that subtracts from `param.grad` is newly reachable
  and was not measured under TP > 1; the benchmark above is tp1.
- `rope_table` is exercised in tests with `use_yarn=False`; the production THD
  path sets it `True`, and the YaRN branch is unchanged from what it replaced.
- The verl example tests now guard with
  `pytest.importorskip("verl", reason=...)`, the idiom four other files in that
  directory already use. A `conftest.py` was tried first and does not work:
  pytest bypasses both `collect_ignore_glob` and `pytest_ignore_collect` for
  paths named explicitly on the command line, which is how the repository's
  runner invokes the suite.
- This branch carries one fix that is not part of the feature. #7064 changed
  `cp_utils.compute_cp_indexer_topk` to return a third value and left lite's
  THD-CP call site unpacking two, so every CP THD forward raised `ValueError:
  too many values to unpack`, on pristine `dev` as well as here. It is included
  because nothing in this branch runs without it; it can be split out on request.
- **Known issue.** This branch's step time varies by 3.47% run to run against
  Core's 0.15% on the same nodes in the same session, with one sample 3% above
  the median. The cause is not established. It is not reflected in the medians
  above, and anyone reading the parity claim should weigh it.

## Issue tracking

Linked issue: 

### Pre-checks

- [x] I have added relevant unit tests — fifteen, in
  `experimental/lite/tests/unit/primitive/test_branch_contracts_unit.py`. Each is
  either a proof that a swapped-in fusion equals what it replaced, or a control
  that must fail for such a proof to mean anything: the wrong rotation, the
  untouched embedding rows, the missing accumulator, the shifted rotary table,
  the head-major grouping.
- [ ] I have added relevant functional tests — none added. No model, config or
  flag is introduced; this is a performance change on an existing path, and the
  end-to-end evidence is the before/after gradient-norm comparison above rather
  than a new functional case.
- [x] I have added proper typing to my code — every new module-level function is
  annotated. The two `torch.autograd.Function` bodies follow the unannotated
  convention Core uses for its own.
- [x] I have added relevant documentation — docstrings on every new symbol,
  including why the rotary cache is keyed the way it is. No user-facing doc
  page: no public API or flag changes.
- [x] I have run `tools/autoformat.sh` — with a caveat worth stating. Its
  pathspec is `megatron/core tests/`, which does not cover `experimental/lite`,
  so it reports an empty changeset for this branch. Ran the same tools by hand
  under the repository's own configuration instead: `black` 26 at
  `line_length = 100` with `--skip-magic-trailing-comma
  --skip-string-normalization`, `isort` 5.13.2, and `ruff check` — all clean on
  the files this branch adds or modifies. Two of them, `csa.py` and `linear.py`,
  are already non-conforming to `black` 26 on pristine `dev`; that is
  pre-existing and left alone rather than folded into this diff.

### CODE REVIEW

**Comment by ISEEKYAN** [comment]
Review evidence for this head.

The vault review panel ran at `70e4b3f45`, before this branch was rebased onto ISEEKYAN dev. That is a different commit from this PR's head `4ea11134`, so the verdict does not transfer on trust — a stale-branch review is a real failure mode, not a formality.

It transfers on evidence: the `experimental/lite` trees of the two commits hash identically.

```
70e4b3f45:experimental/lite  6899a12a467d0a181306d97b2d61327f31e21f4b
4ea11134:experimental/lite   6899a12a467d0a181306d97b2d61327f31e21f4b
```

The rebase moved 22 commits onto a newer dev with no conflicts and no content change under `experimental/lite`. Re-verified on the rebased head anyway: 46 unit tests pass on GPU, and the step time median over three runs is 1372.9 ms against Megatron Core's 1376.3 ms.

Panel verdict at that tree: approve-with-caveats, no blockers. The six caveats are in the release notes section above.

**Comment by ISEEKYAN** [comment]
Correction to my earlier comment: it no longer holds.

That comment showed the review panel's tree and the PR head hashing identically. That was true when I wrote it. It stopped being true when I trimmed the commentary and tests afterwards -- 16 files changed since the reviewed commit, +350/-1089.

```
70e4b3f45:experimental/lite  6899a12a467d0a181306d97b2d61327f31e21f4b   (reviewed)
91ec0e790:experimental/lite  33d6ed733774353cd96388e692e4de1d32b34a5e   (this head)
```

So the approve-with-caveats verdict does not currently cover what is in this PR. A review of a stale tree is exactly the failure mode that evidence was meant to rule out, and I reintroduced it. Re-running the panel on this head now; I will post the result here.

The code paths themselves are unchanged -- the trimming removed prose, deleted two low-value test files and merged others -- and the 46 remaining unit tests still pass on GPU at this head. But that is my assertion, not a review, which is the distinction the earlier comment existed to make.

**Comment by ISEEKYAN** [comment]
Review evidence, closing out my earlier correction.

The panel ran on this head, `1f48d590` — not on an ancestor — so unlike the previous round there is nothing to argue about tree equality this time. Verdict: **approve-with-caveats, no blockers**. Four blind reviewers and four rebuttals raised nothing anchorable that would silently produce wrong numbers or crash at this bar.

The previous round rejected on a real defect of mine: the rotary table cache keyed on the position tensor's address and version counter, which the caching allocator makes unsound — a later tensor of the same shape lands on the same address with its version reset, and the lookup returns tables built for different positions. Wrong RoPE, silently, with every shape and dtype intact.

Four fixes were measured rather than argued:

```
strong reference, keeping the buffer alive     1387.7 ms
weak reference, identity checked on hit        1385.1 ms   (slower than no cache)
no cache at all                                1381.0 ms
tables attached to the packed batch            1366.0 ms
```

The last one is what shipped, and it is not a better cache — it is not an identity cache at all. The tables live on `packed_seq_params`, built once per microbatch and passed to every layer, so its lifetime *is* their validity. No address, no version counter, no liveness trick, and the failure mode does not exist.

That also retracts something: the 1.01% this cache was originally credited with came partly from hits it had no right to. Time saved by being wrong is not a speedup, and the cumulative figure below is recomputed without it.

Where this leaves the branch, against Megatron Core at the same configuration (both arms: indexer loss off, no forced load balancing; medians, noise floor 0.4%):

| | ms/step |
|---|---|
| Megatron Core | 1376.3 |
| lite, before | 1595.3 |
| lite, now | **1366.0** |

Full unit suite in the container: 8 failed / 827 passed, the same eight failures as pristine dev, item for item — no test regressed. End-to-end, gradient norms across the change differ by 7.9e-5 against a same-code re-run spread of 2.09e-4.

The caveats are in the release notes section. Two of them — the `_AllReduceETP` docstring disagreeing with its backward, and the CSA dense-batch refusal not being recorded in the durable unsupported-combinations list — read to me as follow-up work rather than something to fold in here.
